crates/grafbase-sdk/src/jq_selection.rs

```diff
@@ -0,0 +1,137 @@
+    // ┬┴┬┴┤ ͡°)
+    // ┬┴┬┴┤
+    selection_cache: HashTable<usize>,
+    filters: Vec<Filter<Native<Val>>>,
+}
+
+impl Default for JqSelection {
+    fn default() -> Self {
+        Self {
+            arena: Arena::default(),
+            inputs: RcIter::new(core::iter::empty()),
+            selection_cache: HashTable::new(),
+            filters: Vec::new(),
+        }
+    }
+}
+
+impl JqSelection {
+    /// Creates a new instance of [`JqSelection`].
+    ///
+    /// Creates an empty cache of compiled filters.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Selects data from a JSON value using a JQ filter.
+    ///
+    /// This method takes a JQ selection filter string and a JSON value, applies the
+    /// filter, and returns an iterator of the results. The filter is compiled and cached
+    /// for reuse on subsequent calls with the same filter string.
+    pub fn select(
+        &mut self,
+        selection: &str,
+        data: serde_json::Value,
+    ) -> anyhow::Result<impl Iterator<Item = anyhow::Result<serde_json::Value>> + '_> {
+        let hasher = DefaultHashBuilder::default();
+        let hasher = |val: &_| hasher.hash_one(val);
+        let hash = hasher(selection);
+
+        if !self.selection_cache.find(hash, |_| true).is_some() {
+            self.create_filter(selection, hash)?;
+        }
+
+        let filter = self.find_filter(hash).unwrap();
+        let filtered = filter.run((Ctx::new([], &self.inputs), Val::from(data)));
+
+        Ok(filtered.map(|v| match v {
+            Ok(val) => Ok(serde_json::Value::from(val)),
+            Err(e) => Err(anyhow::anyhow!("{e}")),
+        }))
+    }
+
+    fn find_filter(&self, hash: u64) -> Option<&Filter<Native<Val>>> {
+        self.selection_cache
+            .find(hash, |_| true)
+            .and_then(|i| self.filters.get(*i))
+    }
+
+    fn create_filter(&mut self, selection: &str, hash: u64) -> anyhow::Result<()> {
+        let program = File {
+            code: selection,
+            path: (),
+        };
+
+        let loader = Loader::new(jaq_std::defs().chain(jaq_json::defs()));
+
+        let modules = loader.load(&self.arena, program).map_err(|e| {
+            let error = e.first().map(|e| e.0.code).unwrap_or_default();
+            anyhow::anyhow!("The selection is not valid jq syntax: `{error}`")
+        })?;
+
+        let filter = Compiler::default()
+            .with_funs(jaq_std::funs().chain(jaq_json::funs()))
+            .compile(modules)
+            .map_err(|e| {
+                let error = e.first().map(|e| e.0.code).unwrap_or_default();
+                anyhow::anyhow!("The selection is not valid jq syntax: `{error}`")
+            })?;
+
+        self.filters.push(filter);
+
+        let index = self.filters.len() - 1;
+        let hasher = DefaultHashBuilder::default();
+        let hasher = |val: &_| hasher.hash_one(val);
+
+        self.selection_cache.insert_unique(hash, index, hasher);
```

Comment:
Here again you're supposing hash collision don't exist, you must use something like `entry` with `eq` function if you can't guarantee the absence of hash collisions.